### PR TITLE
[Codegen][Common] Resolve `scf.forall` that are used for workgroup distribution

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The `scf.forall` for workgroup distribution need to be resolved after startegy lowering. Since this needs an update to the entry point region, the `scf.forall` resolution is now added with `ReconcileTranslationInfoPass`. If needed this could be moved out of here.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp [hunk 1]
```diff
@@ -32,6 +35,272 @@ class ReconcileTranslationInfoPass final
...
+getMappingPermutation(ArrayRef<IREE::Codegen::WorkgroupMappingAttr> mapping) {
+  return invertPermutationVector(llvm::map_to_vector(mapping, [&](auto a) {
+    return static_cast<int64_t>(mapping.size() - 1) - a.getMappingId();
+  }));
+}
+
+/// Return the procId and nprocs to use for each of the distributed loops,
+/// derived from `hal.interface.workgroup.id/count`s.
+static FailureOr<
+    std::pair<SmallVector<OpFoldResult>, SmallVector<OpFoldResult>>>
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:75]: I think `SmallVector<std::pair<OpFoldResult, OpFoldResult>>` might be cleaner and more explicit that the length of each vector is the same. Also fewer `applyPermutation` calls.

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp [hunk 2]
```diff
@@ -32,6 +35,272 @@ class ReconcileTranslationInfoPass final
...
+  }
+
+  auto inversePermutation = invertPermutationVector(permutation);
+  applyPermutationToVector(procId, inversePermutation);
+  applyPermutationToVector(nprocs, inversePermutation);
+  return std::make_pair(procId, nprocs);
+}
+
+/// Resolve scf.forall operation by using the workgroup ID and counts.
+static LogicalResult resolveForAllOp(RewriterBase &rewriter,
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp]: nit: rename to `resolveWorkgroupForAllOp`

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp [hunk 3]
```diff
@@ -32,6 +35,272 @@ class ReconcileTranslationInfoPass final
...
+      verifyWorkgroupMappingAttrArray(forallOp);
+  if (failed(workgroupMapping)) {
+    return failure();
+  }
+
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(forallOp);
+
+  SmallVector<OpFoldResult> procId;
+  SmallVector<OpFoldResult> nprocs;
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp]: nit: `nprocs` is unused?

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp (hunk 1 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:75]: Edit: I see how this is used and keeping them as separate vectors does seem easier there, leaving this comment as optional.

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp [hunk 4]
```diff
@@ -32,6 +35,272 @@ class ReconcileTranslationInfoPass final
...
+  auto mappingAttr =
+      llvm::map_to_vector(forAllOp.getMapping().value(), [](auto a) {
+        return cast<IREE::Codegen::WorkgroupMappingAttr>(a);
+      });
+  auto permutation = getMappingPermutation(mappingAttr);
+  workgroupCount = applyPermutation(workgroupCount, permutation);
+  return lowerWorkgroupCountFromSliceOp(rewriter, funcOp, workgroupCount);
+}
+
+static LogicalResult resolveForallOps(RewriterBase &rewriter,
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp]: nit: similarly here, `resolveWorkgroupForallOps`

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/ReconcileTranslationInfo.cpp:75]: Thanks for the suggestions. I am going to leave it this way for now.

**Pull request merged**